specs/core/0_beacon-chain.md

```diff
@@ -1285,128 +1271,97 @@ def process_deposit(state: BeaconState,
-    Note that this function mutates ``state``.
-    """
+def initiate_validator_exit(state: BeaconState, index: int) -> None:
     validator = state.validator_registry[index]
-    if validator.status != ACTIVE:
-        return
-
-    validator.status = ACTIVE_PENDING_EXIT
-    validator.latest_status_change_slot = state.slot
+    validator.status_flags |= INITIATED_EXIT
 ```
 
 ```python
-def exit_validator(state: BeaconState,
-                   index: int,
-                   new_status: int) -> None:
-    """
-    Exit the validator with the given ``index``.
-    Note that this function mutates ``state``.
-    """
+def exit_validator(state: BeaconState, index: int) -> None:
     validator = state.validator_registry[index]
-    prev_status = validator.status
 
-    if prev_status == EXITED_WITH_PENALTY:
+    # The following updates only occur if not previous exited
+    if validator.exit_slot <= state.slot + ENTRY_EXIT_DELAY:
         return
 
-    validator.status = new_status
-    validator.latest_status_change_slot = state.slot
-
-    if new_status == EXITED_WITH_PENALTY:
-        state.latest_penalized_exit_balances[state.slot // COLLECTIVE_PENALTY_CALCULATION_PERIOD] += get_effective_balance(state, index)
+    validator.exit_slot = state.slot + ENTRY_EXIT_DELAY
 
-        whistleblower_index = get_beacon_proposer_index(state, state.slot)
-        whistleblower_reward = get_effective_balance(state, index) // WHISTLEBLOWER_REWARD_QUOTIENT
-        state.validator_balances[whistleblower_index] += whistleblower_reward
-        state.validator_balances[index] -= whistleblower_reward
-
-    if prev_status == EXITED_WITHOUT_PENALTY:
-        return
-
-    # The following updates only occur if not previous exited
     state.validator_registry_exit_count += 1
     validator.exit_count = state.validator_registry_exit_count
-    state.validator_registry_delta_chain_tip = get_new_validator_registry_delta_chain_tip(
-        current_validator_registry_delta_chain_tip=state.validator_registry_delta_chain_tip,
-        validator_index=index,
-        pubkey=validator.pubkey,
-        flag=EXIT,
+    state.validator_registry_delta_chain_tip = hash_tree_root(
+        ValidatorRegistryDeltaBlock(
+            current_validator_registry_delta_chain_tip=state.validator_registry_delta_chain_tip,
+            validator_index=index,
+            pubkey=validator.pubkey,
+            slot=validator.exit_slot,
+            flag=EXIT,
+        )
     )
 ```
 
+```python
+def penalize_validator(state: BeaconState, index: int) -> None:
+    exit_validator(state, index)
+    validator = state.validator_registry[index]
+    state.latest_penalized_exit_balances[(state.slot // EPOCH_LENGTH) % LATEST_PENALIZED_EXIT_LENGTH] += get_effective_balance(state, index)
+
+    whistleblower_index = get_beacon_proposer_index(state, state.slot)
+    whistleblower_reward = get_effective_balance(state, index) // WHISTLEBLOWER_REWARD_QUOTIENT
```

Comment:
`get_effective_balance(state, index)` is called twice in this function, could be refactored:
```python
    exit_validator(state, index)
    validator = state.validator_registry[index]

    last_penalized_epoch = (state.slot // EPOCH_LENGTH) % LATEST_PENALIZED_EXIT_LENGTH
    effective_balance = get_effective_balance(state, index)

    state.latest_penalized_exit_balances[last_penalized_epoch] += effective_balance
    whistleblower_index = get_beacon_proposer_index(state, state.slot)
    whistleblower_reward = effective_balance // WHISTLEBLOWER_REWARD_QUOTIENT
```

Reply:
helper var not needed here in the spec imo